### PR TITLE
User inventory rule

### PR DESCRIPTION
**Pull request by Lainow**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !32435

Add new options in rules to manage user reconciliation.
image

### CODE REVIEW

File: src/RuleAsset.php [hunk 1]
```diff
@@ -269,6 +298,54 @@ public function executeActions($output, $params, array $input = [])
...
+                                         $action->fields["value"],
+                                         $regex_result
+                                     );
+                                    if ($res != null) {
+                                          $user = User::getUsersIdByEmails($res);
+                                        if ($user) {
+                                            $output['users_id'] = $user[0];
+                                        } else {
+                                            $output['users_id'] = 0;
+                                        }
```
Comment (stonebuzz) [inline src/RuleAsset.php:313]: we should not reset the field to 0

If the correct user is not found, leave the field untouched.

File: src/RuleAsset.php [hunk 2]
```diff
@@ -269,6 +298,54 @@ public function executeActions($output, $params, array $input = [])
...
+                                         $action->fields["value"],
+                                         $regex_result
+                                     );
+                                    if ($res != null) {
+                                          $user = User::getIdByField('registration_number', $res);
+                                        if ($user) {
+                                            $output['users_id'] = $user;
+                                        } else {
+                                            $output['users_id'] = 0;
+                                        }
```
Comment (stonebuzz) [inline src/RuleAsset.php:329]: we should not reset the field to 0

If the correct user is not found, leave the field untouched.

File: src/RuleAsset.php [hunk 3]
```diff
@@ -269,6 +298,54 @@ public function executeActions($output, $params, array $input = [])
...
+                                            $action->fields["value"],
+                                            $regex_result
+                                        );
+                                    if ($res != null) {
+                                            $user = User::getIdByField('sync_field', $res);
+                                        if ($user) {
+                                            $output['users_id'] = $user;
+                                        } else {
+                                            $output['users_id'] = 0;
+                                        }
```
Comment (stonebuzz) [inline src/RuleAsset.php:345]: we should not reset the field to 0

If the correct user is not found, leave the field untouched.

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: tests/functional/Glpi/Inventory/Inventory.php [hunk 4]
```diff
@@ -7851,4 +7852,153 @@ public function testStatusIfInventoryOnAddUpdate()
...
+            <NAME>glpixps</NAME>
+            <UUID>25C1BB60-5BCB-11D9-B18F-5404A6A534C4</UUID>
+          </HARDWARE>
+          <BIOS>
+            <MSN>640HP72</MSN>
+          </BIOS>
+          <USERS>
+            <DOMAIN>TECLIB</DOMAIN>
+            <LOGIN>" . $xml_fields['name'] . "</LOGIN>
+          </USERS>
```
Comment (stonebuzz) [inline tests/functional/Glpi/Inventory/Inventory.php:7983]: Can you add another couple of data (```<USER>``` / ```<DOMAIN>```) to test with another criteria like ```Rule::PATTERN_NOT_CONTAIN``` 

```xml
          <USERS>
            <DOMAIN>TECLIB</DOMAIN>
            <LOGIN>" . $xml_fields['name'] . "</LOGIN>
          </USERS>
          <USERS>
            <DOMAIN>TECLIB</DOMAIN>
            <LOGIN>Admin</LOGIN>
          </USERS>
```
Criteria => ```Rule::PATTERN_NOT_CONTAIN``` ```Admin```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/RuleAsset.php [hunk 5]
```diff
@@ -269,6 +298,48 @@ public function executeActions($output, $params, array $input = [])
...
+                                     $res = RuleAction::getRegexResultById(
+                                         $action->fields["value"],
+                                         $regex_result
+                                     );
+                                    if ($res != null) {
+                                          $user = User::getUsersIdByEmails($res);
+                                        if ($user) {
+                                            $output['users_id'] = $user[0];
+                                        }
+                                    }
```
Comment (stonebuzz) [inline src/RuleAsset.php:312]:
```suggestion
                                          $user = User::getFromDBbyEmail($res);
                                        if ($user) {
                                            $output['users_id'] = $user->fields['id'];
                                        }
                                    }
```

**Review by stonebuzz — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by stonebuzz** [comment]
waiting for customer feedback

File: src/RuleAsset.php [hunk 6]
```diff
@@ -269,6 +298,49 @@ public function executeActions($output, $params, array $input = [])
...
+                            case "_affect_user_by_email_and_regex":
+                                foreach ($this->regex_results as $regex_result) {
+                                     $res = RuleAction::getRegexResultById(
+                                         $action->fields["value"],
+                                         $regex_result
+                                     );
+                                    if ($res != null) {
+                                        $user = new User();
+                                        $user->getFromDBbyEmail($res);
+                                        if (isset($user->fields['id']) && $user->fields['id'] != 0) {
```
Comment (cedric-anne) [inline src/RuleAsset.php:310]:
```suggestion
                                        if ($user->getFromDBbyEmail($res)) {
```

File: src/RuleAsset.php [hunk 7]
```diff
@@ -269,6 +298,49 @@ public function executeActions($output, $params, array $input = [])
...
+                                            $regex_result
+                                        );
+                                    if ($res != null) {
+                                            $user = User::getIdByField('sync_field', $res);
+                                        if ($user) {
+                                            $output['users_id'] = $user;
+                                        }
+                                    }
+                                }
+                                break;
```
Comment (cedric-anne) [inline src/RuleAsset.php:343]: Please fix the indentation on these lines.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**